### ISSUE TITLE
Network applet: Show connection status

### ISSUE THREAD

**Issue by danboid** [SerenityOS/serenity#6062]
Currently the network applet always shows the same icon whether the nic is connected or not. It would be better if it showed a different icon when there is no connection and the mouseover popup displayed whether you are connected or not.

**Comment by bcoles** [comment]
> 
> 
> Currently the network applet always shows the same icon whether the nic is connected or not. It would be better if it showed a different icon when there is no connection and the mouseover popup displayed whether you are connected or not.

This is already implemented.

**Comment by bcoles** [comment]
Both of these suggestions have existed since the Network applet was introduced in #4461.

The icon changes when the network is down. You can test the network down state by setting an invalid IP address for the adapter; ie:

```
# ifconfig -a e1k0 -i 0.0.0.0
```

Or simply boot on a system for which we don't have network drivers.

When you mouse over the network applet a popup is presented showing the IP address for all interfaces (except loopback).

Of course, whether the interface has an IP address has nothing to do with whether the network is "connected", as described here https://github.com/SerenityOS/serenity/pull/4461#issuecomment-748467818.

Ideally the applet should both and IP address inbound network traffic packet count to determine whether a connection is live.

**Comment by bcoles** [comment]
A better title for this issue would be something like: "Network Applet: connection status should use packet count"

Implementing the correct connection status detection and associated display icon depends on how we want to handle multiple network adapters. https://github.com/SerenityOS/serenity/pull/4461#issuecomment-748467818

Systems with more than one network adapter are common. Even on a system which has only one Etherner network adapter (1 NIC) and a loopback adapter (1 NIC), it may also have a WiFi adapter (1 NIC), a USB Ethernet adapter (1 NIC), virtualization software (VMware/VirtualBox) installed (2 NICs), one ore more VPN clients installed (1+ NICs), occasionally uses a tethered 5G phone/modem (1 NIC), has development drivers (such as WinPCap) installed (1 NIC), ...

We don't support most of that functionality yet (we do support multiple NICs and will likely eventually support WIFi), but it should be taken into consideration for future.

The default approach on modern Windows systems is to show the network icon as connected so long as at least one network interface has a connection. However, I don't like this approach, as the entire purpose of the network icon is to act as a visual indicator as to network status. Hiding status for some adapters undermines the entire purpose of the icon (in instances where more than one network interface is in use).

I prefer the traditional approach on Windows 2000 era of showing multiple network icons (one for each adapter). But this can become cluttered rather quickly. A workaround is to allow users to configure whether a network icon is displayed in the taskbar on a per-adapter basis:

* Always hidden
* Always shown
* Hidden when disconnected / Shown only when connected

**Comment by danboid** [comment]
OK so this is a bug report and not a feature request. Its a bug report because:

* On my bare metal serenity box with 1 RTL NIC (I don't count loopback as a nic for these purposes), the icon doesn't change when I physically remove the cable

* The icon shows as connected even though it actually isn't. The system monitor shows packets are being received but they get discarded because its not configured correctly (wrong IP and/or subnet)

Loopback should be excluded from having an icon IMO but like @bcoles I would prefer one icon for each nic.

**Comment by bcoles** [comment]
> On my bare metal serenity box with 1 rtl8139 nic (I don't count loopback as a nic for these purposes), the icon doesn't change when I physically remove the cable

Correct. The icon only changes when the IP address changes (including when it is changed to null).

Does the IP address change when you disconnect the network cable? I suspect that it does not. At least, probably not until the DHCP lease expires which will take a while.

If the IP address doesn't change (to null) then there's not much that Network applet can do about it. The applet relies on the kernel being truthful. The kernel or networking service should track changes to the IP address and adapter state.

> The icon shows as connected even though it actually isn't.

Correct. The icon only changes when the IP address changes (including when it is changed to null).

This could be augmented to also check for incoming network traffic. It could also be modified to check an external host for connectivity. This approach is taken by modern systems to check whether a network has a captive portal. However, I don't like this approach of pro-actively calling out to a host on the Internet.

> The system monitor shows packets are being received but they get discarded because its not configured correctly (wrong IP and/or subnet) hence I don't like your suggested title of "Network Applet: connection status should use packet count" as that wouldn't fix this issue without the config being took into consideration too,

What do you suggest calling this issue? The current title `Network applet: Show connection status` is a useless out-of-context assertion, like a homeless man yelling at cars on the highway. It doesn't describe what you're experiencing nor does it describe a bug.

> Loopback should be excluded from having an icon IMO but like @bcoles I would prefer one icon for each supported non-loopback nic.

Agreed. The loopback interface should not have an icon. Currently the loopback interface it is excluded from determining whether the network is connected by default when displaying the icon.

https://github.com/SerenityOS/serenity/blob/563712abcef5ca791bdbc09c4a02931fd709c7c9/Userland/MenuApplets/Network/main.cpp#L126

**Comment by bcoles** [comment]
Suggest something like: `Network Applet: Only IP address is used to check network connection status`

**Comment by danboid** [comment]
"Does the IP address change when you disconnect the network cable? I suspect that it does not. At least, probably not until the DHCP lease expires which will take a while."

No, it doesn't lose the incorrectly assigned address when I remove the cable.

**Comment by bcoles** [comment]
> No, it doesn't lose the incorrectly assigned address when I remove the cable.

Incorrectly assigned by DHCP? That's unusual.

> Why is it hellbent on assigning itself a 10.x.x.x address?

Unusual that you should receive an incorrect DHCP lease on a real network.

Are you getting `10.0.2.15` ?

Traditionally most development has been done on QEMU which defaults to `10.0.2.15` for the emulated adapter inside Serenity. If there is a bug which forces a `10.x.x.x` address it may have been missed, but there's no immediately obvious reason why this would be the case.

The address isn't hard-coded anywhere except `Meta/run.sh` and there's no fallback default IP address.

**Comment by danboid** [comment]
Yes, 10.0.2.15. See https://github.com/SerenityOS/serenity/issues/4412

**Comment by alimpfard** [comment]
@bcoles it is hardcoded, just in an unsearchable way
https://github.com/SerenityOS/serenity/blob/master/Kernel/Net/NetworkTask.cpp#L71

That's normally just fine, but when the dhcp client doesn't work for one reason or another, it's no longer fine :P

**Comment by bcoles** [comment]
> 
> 
> @bcoles it is hardcoded, just in an unsearchable way

and not in `DHCPClient`. and not in `Userland/Services`. lol

> That's normally just fine, but when the dhcp client doesn't work for one reason or another, it's no longer fine :P

Seems like it is an unreasonable default? Especially now that we have a DHCP client.

It is safe to remove in QEMU. Perhaps it should be removed?

```diff
$ gitd Kernel/Net/NetworkTask.cpp
diff --git a/Kernel/Net/NetworkTask.cpp b/Kernel/Net/NetworkTask.cpp
index 868997cc7..6a325a8ee 100644
--- a/Kernel/Net/NetworkTask.cpp
+++ b/Kernel/Net/NetworkTask.cpp
@@ -60,26 +60,16 @@ void NetworkTask::spawn()
 void NetworkTask_main(void*)
 {
     WaitQueue packet_wait_queue;
-    u8 octet = 15;
     int pending_packets = 0;
     NetworkAdapter::for_each([&](auto& adapter) {
+        dmesgln("NetworkTask: {} network adapter found: hw={}", adapter.class_name(), adapter.mac_address().to_string());
+
         if (String(adapter.class_name()) == "LoopbackAdapter") {
             adapter.set_ipv4_address({ 127, 0, 0, 1 });
             adapter.set_ipv4_netmask({ 255, 0, 0, 0 });
             adapter.set_ipv4_gateway({ 0, 0, 0, 0 });
-        } else {
-            adapter.set_ipv4_address({ 10, 0, 2, octet++ });
-            adapter.set_ipv4_netmask({ 255, 255, 255, 0 });
-            adapter.set_ipv4_gateway({ 10, 0, 2, 2 });
         }
 
-        dmesgln("NetworkTask: {} network adapter found: hw={}, ipv4_address={}, ipv4_netmask={}, ipv4_gateway={}",
-            adapter.class_name(),
-            adapter.mac_address().to_string(),
-            adapter.ipv4_address(),
-            adapter.ipv4_netmask(),
-            adapter.ipv4_gateway());
-
         adapter.on_receive = [&]() {
             pending_packets++;
             packet_wait_queue.wake_all();
```

**Comment by alimpfard** [comment]
> Perhaps it should be removed?

Yeah, I don't think we have anything that requires an IP address to be set in early boot, or before DHCP does its thing, so it's probably not useful :shrug:

**Comment by danboid** [comment]
I know I'm not the only person to have tried networking on bare metal so I'm surprised this wasn't surfaced already.

**Comment by panos** [comment]
Leaving a note here and cross-linking to #6447 (This would affect `IF_UP` and `IFF_RUNNING`, the latter of which has not been included yet.)
